### DISCUSSION TITLE
Remove empty tags in XML Object?

### DISCUSSION THREAD

**Discussion by elemle** [nextgenhealthcare/connect#4819] [Q&A] [answered]
Remove empty tags in XML Object?
Is there a way to do this with one for loop and maybe using the .descendants() instead of .children() in E4X?
(The code below basically removes empty XML tags for 2 levels deep.)
//-----------------------------------------------------------------
  for each (var level1 in xml.children()) {	// descendants?  	
  	if(!level1.toString() || level1.toString().trim()=="") {	
  		delete (xml.children()[level1.childIndex()]);
  	}
  	else {
  		for each (var level2 in level1.children()) {	
  		  if(!level2.toString() || level2.toString().trim()=="") { 
  		    delete (level1.children()[level2.childIndex()]);
  	          }
  	        }		
  	}
  }
//------------------------------------------------------------------
Mirth Connect Server 3.11.0
Built on April 6, 2021
Java version: 1.8.0_181

**Comment by rjkroll** [comment]
If I am generating XML, I have a final transformer step that does the following to remove empty nodes/tags:

Type: XSLT
Source XML String:  xsltTransformedMsg
Transfromer Factory: Default
XSLT Template:

<xsl:stylesheet version="1.0" xmlns:xsl="http://www.w3.org/1999/XSL/Transform">
  <xsl:strip-space elements="*"/>
  <xsl:output indent="yes" />
  <xsl:template match="@*|node()">
    <xsl:if test=". != ''">
      <xsl:copy>
        <xsl:apply-templates  select="@*|node()"/>
      </xsl:copy>
    </xsl:if>
  </xsl:template>
</xsl:stylesheet>



Then, in the Destination Template, I use the variable defined above :   ${xsltTransformedMsg}

**Reply by elemle** [reply]
Nice, but not the droids I am looking for.

**Reply by pacmano1** [reply]
Huh?  What wrong with using an XSLT transform to achieve your objective?

**Reply by elemle** [reply]
I am asking if there is javascript that I can write that goes through all the descendants instead of only one level of children.

**Comment by pacmano1** [comment]
I would do what @rjkroll suggested, but use different XSLT

![image](https://user-images.githubusercontent.com/44065187/139339770-87087d83-798c-4650-a1bb-a42e76b675dc.png)

Before:

![image](https://user-images.githubusercontent.com/44065187/139339833-7a8976dc-fdb9-4d5d-982a-18c323acf4e1.png)

After:

![image](https://user-images.githubusercontent.com/44065187/139339955-9add5a15-bfc8-463f-bb92-9136f1f93113.png)

**Reply by pacmano1** [reply]
Also see https://github.com/nextgenhealthcare/connect-examples/tree/master/Code%20Templates/Strip%20Empty%20Nodes%20from%20XML

**Reply by elemle** [reply]
Awesome.  That is exactly what I was looking for...  a javascript code template to do this.  Thanks for the narupley code!

**Comment by pacmano1** [comment, marked as answer]
Answered, see thread
